### ISSUE TITLE
Crash on leto_groupby

### ISSUE THREAD

**Issue by streamflex** [elchs/LetoDBf#17]
A compound expression as the grouping parameter to leto_groupby crashes the server. Logs show (for simple test program groupby):

01.16.2019 17:07:07 INFO: connected  127.0.0.1:58563 groupby CP: EN  DF: dd.mm.yyyy  conn-ID 0
01.16.2019 17:07:07 !!!!! leto_errInternal !!!!!

Request was:

aResult := cst->(leto_groupby("currency+payterms","#,crlimit,duerem","inv1line"))

where cst is the dbf alias, currency,payterms,crlimit,duerem & invline are valid fields. This actually produces what appears to be correct output before the server dies. Simplifying the group expression to a single field produces correct output and avoids the server crash.

Perhaps compound expressions for the grouping field are not supported by design. But I'd rather the call failed than crashed the server! It would be very cool to support compound expressions, because then I could group by any combination of fields.

Note that leto_dbeval can be forced to do the same job correctly, using a rather ugly long codeblock:

  cExec := "{|n,x| if(n==1,x:={},nil),n:=ascan(x,{|v|v[1]==currency+'|'+payterms}), if(n==0,aadd(x,{currency+'|'+payterms,1,crlimit}),x[n][2]++),if(n==0,nil,x[n][3]+=crlimit),x}"
   aResult := cst->(leto_dbeval(cExec,"inv1line"))

**Comment by elchs** [comment]
Hi,
for very sure a server crash is an *absolutely* must to fix.

leto_GroupBy() at server side is a 400 lines source block, [inherited and not well tested],
but as you got correct result, there should be only few lines left i have to inspect.
I will comment again if more details are needed  ...

best regards
Rolf

**Comment by streamflex** [comment]
Thanks. Thinking further about this function, I wonder whether it could all be generalised through use of UDFs in leto_dbeval? A range of UDFs could be provided for different grouping features. For example, I've set up a UDF to calculate multiple group keys, with the statistical functions {sum, min, max} for each requested value. That could easily be extended to mean and variance.

   cExec := "{|n,x| udf_GroupCalcHash(n,x,{currency,payterms},{crlimit})}"
   aResult := cst->(leto_dbeval(cExec,"inv1line"))

The input arrays are group keys and calculated values respectively. It returns a hashmap (wrapped in a single element array, because array seems to be the only acceptable return value).

If a set of suitable functions is provided in letoudf.prg, then this would be an 'out of the box' feature, but very easily extensible too. Then leto_groupby() and its 400 lines of mind-bending C code could probably be deprecated :). Eg, there could be udf_GroupBySum(), udf_GroupByStats(), udf_GroupByStatsExt().

If this looks interesting then I'll give you a set of UDFs for grouping, to use whatever way you see fit.

**Comment by elchs** [comment]
Hi,

cause of crash fixed, please counter-check.

The 'compound expression' for grouping surely can utilize also Harbour functions,
all these listed as "#REQUEST" in: source/server/server.prg.
This is true also for the 'fields" columns.

Function 'Leto_GroupBy()' is, as said, inherited from origin LetoDB,
and should fulfill its designed purpose -- will keep it same as function 'Leto_Sum()',
so no need to hide it behind 'Leto_DbEval()'

Function 'Leto_DbEval()' is extensively tested and can do more than 'DbEval(),
for which it's a drop-in replace.
It's for performing miracles lightning fast at server-side, call it a blockbuster ;-)

best regards
Rolf

**Comment by streamflex** [comment]
Thanks for speedy fix Rolf, much appreciated. The leto_GroupBy call now both shows expected results and the server does not crash.

I agree, leto_dbeval is a great feature. As described above I can now effectively reproduce SQL 'group by' functionality using it. Let me know if you're interested in any extensions like this.

One further thing is worth mentioning. Requesting additional Harbour functions such as the hashmap functions means patching server.prg. Next time the source code is pulled these changes must be backed out and manually reapplied before building. Also, only letoudf.hrb appears to be recognised.

I can't think of a fix for the symbol request issue. Requesting in the UDF file (and thus in the HRB) does not work.

However a simple extension to server.prg will pick up any .hrb file in the bin directory.
`STATIC PROCEDURE loadHrbFiles()
   local aDir
   aDir := directory(s_cDirBase +"*.hrb")
   if empty(aDir)
     return                                    // nothing found
   endif
   aeval(aDir, {|f| leto_hrbLoad(f[F_NAME])})
   return
`
and calling this rather than leto_HrbLoad() directly. This does not work for the 'reload' command, but that's easy enough to improve. Picking up user-defined UDF packages separately from letoudf.hrb allows a more modular approach and avoids user extensions being overwritten when pulling from the LetoDBf repo.

**Comment by elchs** [comment]
Hi,

so i will 'close' the 'crash issue' -- that shouldn't hinder us of further discussions in this thread.

Convince! me ;-) of missing Harbour functions to be available by default at server,
highly wanted by many! users -- then i add them as `#REQUEST` in 'server.prg'.
So far i like to have a "light" server, only 'spoiled' with really essential functions.
Aside that, look into 'letodb.hbp' for: LETO_FULLCMDSET_[HB|CT],
its just out-commented by default ...

When the server is starting, it only looks by default for "letoudf.hrb", and by default none is there.
To use the server in production, you copy the executable out of the 'bin' folder to its 'final place',
and there you then can have a personal "letoudf.hrb".
As i explained recently at Harbour user forum, by intention! no existing Harbour function is overloaded,
especially not these introduced by LetoDBf server.
[ BTW, i more regularly look there for posts as here at GitHub, so if you need urgent bug-fix .. ]
Further i wouldn't be able to play 'blindfolded chess' to answer questions, it must behave 'predictable'.

---
SQL: i would like the LetoDBf server to (limited) understanding SQL queries,
and i was experimenting with the Harbour pre-processor to analyze a SQL query,
and to translate it into a 'Leto_DbEval()' call.
I won't explain that further, it's still FAR from ready even i invested quite a lot of time.

best regards
Rolf

**Comment by streamflex** [comment]
Hi Rolf,
Clearly I can't speak for other users :). The hb_hX() functions are valuable for processing group results for large datasets. An ascan() works fine on the scale of a few thousand rows. But its performance will degrade exponentially as the number of distinct group keys grows.

Applying my own 'request' statement in server.prg each time I pull a new release is not the end of the world. Neither is restricting the loaded HRB to letoudf.hrb. Neither for that matter is having group functions as my own UDFs. It all works.

However, as a product design question, addressing these issues doesn't look difficult to implement and IMO provides a big step upwards in capability and ease of use. My main thought coming to Harbour is how much detailed technical knowledge is needed to use it effectively. It's been designed by very technical people for other very technical people. Better documentation and a friendlier face would help it to become more popular. Most good programmers can work out what they need given time, but how many are prepared to devote the time?

Having played with leto_dbeval for grouping, I am pleased with the results. I have included it into a general query management framework, and can drop in any set of group key and measure expressions using very simple code. The following improvements would make it very good to use:

1. Return a 'flat' array of results, even though the intermediate calculations use a hashmap. This avoids re-processing the result set client-side to convert from concatenated hash key plus array of values into a flat array, which can then be used onward for browsing, etc. This work could easily be done server-side if a separate UDF is passed for post-processing
2. Provide wrapper functions for leto_dbeval plus UDFs. Something like leto_Group(aKeyExpr,aMeasureExpr,nStats,cFor), where aKeyExpr is string expressions for group keys, aMeasureExpr is string expressions for group measures, nStats is the type of statistics which will be calculated for each measure. Eg, sum only, sum/min/max, sum/min/max/mean/var, etc. For ultimate flexibility nStats could become aStats == parallel array to the measures array with the type of statistic functions for each measure.
3. The row count would always be returned, so no need to request it specially.
4. Currently the leto_dbeval() cFor expression does not accept elements from related work areas. It quietly returns nothing. That's inconsistent with the server-side dbSetFilter(), which does seem to accept any valid expression.

Regarding SQL, my current UDF-based implementation of leto_dbeval is actually quite close to something that I think could be added to the command processor:

select <measure1>[,...<measureN>]
from <dbf>
where <cFor clause>
group by <groupkey1>[,...<groupkeyN>]

translates more or less directly to:
cExec := "{|n,x| udf_GroupCalcHash(n,x,{%GB%},{%GM%})}"
cGroupKey := "{groupkey1,groupkeyN}"
cMeasure := "{measure1,measureN}"
cExec := strtran(cExec, "%GB%", cGroupKey)
cExec := strtran(cExec, "%GM%", cMeasure)
aResult := (dbf)->(leto_dbeval(cExec, cFor))

We could also consider joins, having and order clauses.

Joins should be relatively easy, translating to dbSetRelation() calls.

Having would be a post-processing task, again through the UDF. The trick would be to insist (much like regular SQL) that the having clauses exactly match one of the measure or group key clauses. The main problem relates to the 'nStats' feature. Without allowing explicit statistical functions to be requested for measures, a having clause would have to support a restricted vocabulary of functions. Perhaps the simplest idea would be to support such a vocabulary for both measure and having clauses. Requested functions would build the proposed aStats parameter. I'm not sure how the command processor could be used for such vocabularies.

I suggest that the order clause is not supported and that ordering will always be by ascending group key values. I suppose ordering could be supported by the post-processing UDF simply as an asort on the final result set...

As discussed above, an important design choice is whether to use the 'mode' suggestion, or to allow individual statistical functions to be requested. The latter would involve calculating all statistics and filtering them to the requested set in the post-processing step mentioned above.

Depending what you think of all this, I could help implement quite a lot of it.

**Comment by elchs** [comment]
Hi Peter ?,

as Thanks for finding a (rare) cause for a server crash,
your wanted HB_H*() functions are added, disabled by default.
See last ChangeLog.txt entry at top, how personally to activate.

---
Hashes similar to arrays are based on an Harbour 'item' at C-level,
look at them as an two-dimensional array as you can write:
aHash[ x ] := "elch"
Working with them is about ( half time ?) slower ! as with an one-dimensional array,
where they surely are convenient to handle 'named array items'.
At my old development machine 1 million hash values add in ~ 0.6 s,
same amount array items in ~ 0.3 s

Hash(-arrays) can be 'flattened' by using: hb_HKeys() or hb_HValues()

https://groups.google.com/forum/#!forum/harbour-users

---
No !, a RELATION to a field in 'cFor'-condition is not a problem for Leto_DbEval()
-- problem for sure hides behind a somehow wrong usage.
Even i can say that 'without tests', i did a test.
Ensure also in letodb.ini: No_Save_WA = 1

Fields with ALIAS in FILTER conditions is a very different complex topic.
FILTER should be NOT set at server side, but at client who informs server.
FILTER expressions not executable at server ( local var, ...) are handled (slow) locally.

When applying a filter in a letoudf.hrb function, it must be re-set with function ending,
as moreover the whole environment have to be restored.
** the client side doesn't know anything about an UDF have done at server **
See for usage of: leto_SetEnv() and leto_ClearEnv()

It looks to me, you start with one of the most complex parts: UDF,
without good experience how the client-server works ?
And you are a bit excused, i should write a bit in Readme.txt ...

LetoDBf is used as RDD: Replaceable Database Driver,
means applications formerly using DBFNTX/CDX just switch to LETO.
All what you seem to do with the supposed wrappers, can be done from client side,
and then also with other RDDs.
Check for meaning of parameters: lRest & lStay ( lStay LETO specific) , that in combination for
cFor and cWhile is grouping, aka multiple calls for a sorted table from client side.
Maybe look also into: include/leto_std.ch for some more examples,
there used for pre-processor command translations like AVERAGE, SUM, ...

As i see me myself as 'system-developer', i create the base,
on which someone can put on his special needs.
After creating the possibility to load an letoudf.hrb my support normally ends,
what's in there is commonly not of my interest.

----
You called LetpDBf a "product" -- phew!, it implies i have got a very single $ for my piece of art!
As not, i have the freedom to decide what lastly is 'painted',
But i do that carefully and in best manner,
i.e. i don't know of even a single unmaintained BYTE [ memory leak ]

BTW, LetoDBf is not the result of 'some weeks' -- think about netto! 3 years,
it started (secretly;) long before the initial upload to GitHub.


---
About: " .. technical people for other very technical people.",
correct: .. very technical people for other technicians ..
Simplified can be said, that not everybody must create software for others,
and there seem fewer who make software [or OS] for other developers.

As invited Harbour core-team member you may expect some extended insights in Harbour,
else i wouldn't be able to create LetoDBf.

---
SQL:
only a bit "group-by" is not what i think of what SQL represents.
Cite:
" .. Joins should be relatively easy, translating to dbSetRelation() calls. "
A relational DBMS means: one record points to another record,
where a SQL JOIN points to a group/set of records.
Such can be IMO implemented by *complex* skip-ing in the relationed workarea,
where there are left/ right/ inner and cross JOINS.
[ work about already in LetoDBf ]

Further a FROM can not only refer to multiple tables, but to another SELECT,
to be implemented as 'temporary' table.
[ LetoDBf already can handle temporary tables, gone again by closing them ]

I fear you didn't got my point:
a text-ual written SQL query send to server,
resulting in a series of DBF actions, to return a result-set "INTO" e.g. a variable/ hash-table.
* Not with one foot hopping around the table imitating a kangaroo * :-)
Such job would start with a query analyze about whats in there,
then open/close tables with wanted ALIAS, creating temporary index etc etc ...,
and in between is the blockbuster Leto_DbEvil()

In other words: like to re-invent an expensive SQL ODBC driver for DBF tables,
then build into the server.
Or likely what Advantage (ADS) server does, do you know that PRODUCT! ?

To repeat from last post:
I won't explain that further, it's still FAR from ready.
So far thanks! for your offer to help -- if once 3/4 ready, when i really get it so far, you can help testing,
working at C-level is commonly reserved to me.

---
BTW, for what does this "sapo" ?
( the Cl*pper etc in there should be removed, not 'distributed' ... )

best regards
Rolf

**Comment by streamflex** [comment]
Hi Rolf,
Ok understood, let's leave it for now.

The purpose of a SQL-like interface would just be ease of use. Rather than the developer worrying about open and closing tables, relations, scopes, record pointers, work areas, etc, he would just issue a query and consume the results. But perhaps that's extending the model further than makes sense. After all this is an xBase tool.

I have removed the sapo2 repository. It should be private and has been hosted elsewhere for some time.

Thanks for your great work on this project, it's much appreciated!
Peter

**Comment by elchs** [comment]
Hi Peter,

you just did another good thing for all LetoDBf fans:
Leto_SetEnv() and Leto_ClearEnv() need an urgent 'face-lift' :-),
to be capable to simultaneous work with *multiple* workarea,
and to be more convenient.
Expect soon a new upload about ...

Basically it would be possible to do that automatized, what these both functions do.
But then it would need to iterate through *all* opened workarea,
one time before the UDF-function is executed and one time afterwards,
where opened workarea can be quite a lot.
Theoretical limit is 65534, without deep explanation why, Harbour practical limit is about 60000,
and 100+ are not uncommon for a big application.
It would result in quite a lot of action, for maybe one or even none changed workarea.
Solution could be a 'flag' for Leto_UDF(), but any way the developer have to take care about,
as working with data changing UDF - functions does in general ...

---
P.S.
had downloaded 'SAPO', looked into this *bunch* of well formatted source.
Respect!, very *well done* -- a very technician at work, a pleasure to look at.
Did you create that ? -- and i still would like to know what it does ?

best regards
Rolf

**Comment by streamflex** [comment]
Hi Rolf,
Thanks for kind words about SAPO. Actually I felt the code needed a strong refactor to make it better structured :). As you can tell it is a very old application, which I have been asked to bring up to date in Harbour. It's an ERP solution for paper mills, but has a very strong generic framework which could be used for any type of database application. I originally designed it and led the development team. Over the past year I've rewritten the most interesting of the modules (Statistical Quality Control) to work wholly with MySQL. The rest will remain as DBFs using LetoDBf. It has been a great pleasure to work with Harbour (apart from the documentation), but this is a spare time project and I will not have time to fully bring it up to the level that I know could be possible.

I have considered opening up the framework as an open source project, but honestly I don't have the time to do it justice. My day job is far too demanding.

Your ideas on Leto_SetEnv() look like a good step forward. Taking a further step, it would be very powerful for server-side functions to simply set up and destroy the necessary server work area context without worrying about the current client-side context. That way the user can mix the familiar client-side RDD API for manually manipulating data row by row, together with set-based query functions (and possibly also set-based CRUD operations). The two paradigms would not be entangled, which hopefully would reduce application complexity.

I can already think of one technical objection to that concept, but hope that it could be overcome without too much trouble.

As you say, much of the needed work to make more sophisticated server-side query features could be done efficiently and cleanly in Harbour code, if there could be a pre-processing and post-processing step built into the leto_dbeval iterator. Then many tasks could be done server-side using 'UDF's. Many (most?) users will be unclear about the precise split between client and server. I suspect many would welcome a more 'packaged' approach with a library for server-side querying that takes care of the housekeeping while presenting a simple interface. Client-side state would be unchanged by server-side functions. Perhaps simulating SQL using the command pre-processor is overkill, but the principle is the same for function-based code.

One further thought - I could possibly help you with documentation. As a native English speaker, naturally I don't fluently speak any other language :(. But I could help with the English version, which might in turn help others to translate to their language.

Peter

**Comment by elchs** [comment]
Hi Peter,

there are two very! different server modes [ No_Save_WA option ],
and further for both the config option: logical versus physical locking.
Using data collecting/ *modifying* UDF at server, it best be configured:
No_Save_WA = 1
Share_Tables = 1
Above i can't (and don't want) to force, as it's possible for other modes.
( Background: No_Save_WA = 0 is inherited & fixed, == 1 my work )

Limits and convention should be described, work should be convenient,
but i never will be able to make that 100% 'fool-proof'.
That means someone writing UDF, especially such dealing with table data,
needs basic knowledge about internals, but can't be stopped doing nonsense.

UDF functions at server are traditionally a single function/ small snippet,
which fill a 'gap'. Think of a complex formula expressed with Harbour command,
to be used in an index expression.
This case it would be not even executed directly by client,
but server need to know it when the index order is created/ updated.

Extending this to a 'group of functions' [mini-application],
i would name it: remote code execution.
Sure you know the 'stored procedure' concept in SQL server,
at end it moves client logic to the server, which is a design question.
For huge tasks we can even create a Harbour application running at server,
as it can co-work with LetoDBf on same DBF tables (in correct configuration)

And moreover the client work can be heavy tuned, depending on situation.
See 'tests/test_mem.prg', which deals about "transactions", skipbuffer-size, tables in server RAM, ...
The `#defines` at top configure it to behave different:
did you know, it's possible to append **many** K ! records a second ?

---
New version of: Leto_SetEnv() and Leto_ClearEnv() uploaded,
and i will look on occasion, if more can be automated ensured by my side.

For sure i still miss something in new remarks for chapter: 9.
BTW: when much too much must be explained,
by experience it commonly indicates some work is missing :-)


---
Yeah, a face-lift for the Read-me -- from pure text to 'markdown',
keeping *basic* readability in a pure text editor,
but be nice displayable and handy (with internal links) in an viewer.
And then it can be converted automated even further to HTML.
I'm just experimenting with this markdown, looks easy but is new to me.

Above alongside improving 'a bit' ;-) my not native Brits,
which a native of similar dialect hopefully understand,
but i.e. Groggle often have 'problems' in automated translation.
( know that of Harbour forum, when i translate my written words back to motherlanguage )
Proofreading by someone implies he understood the topic
-- please let me know of fully not-understandable chapters!

Unsolved problem of an automated translation is,
that `keywords` like config-options/ functions & file names must stay origin,
and only the description to be translated.
Herefore i found an HTML tag, but it finally would remove readability in a text editor.
Tips welcome!

---
What you can do best for LetoDBf:
try it even at its edges -- and file issues if found.

very best regards
Rolf
